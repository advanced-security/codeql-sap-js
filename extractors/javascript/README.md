# Extension to the JavaScript extractor to support CDS compilation

This directory contains a `pre-finalize.sh` script that can be dropped into the `tools` directory
of a JavaScript extractor to enable the automatic compilation and extraction of SAP CAP CDS files
during the `codeql database create` process.

The script requires that the `cds` extractor is available when `codeql database init` is called. This can either be through providing the extractors parent directory using the `--search-path` bundle, or by including the extractor (and this pre-finalize script) in a custom bundle created by the [CodeQL Development Toolkit (qlt)](https://github.com/advanced-security/codeql-development-toolkit).