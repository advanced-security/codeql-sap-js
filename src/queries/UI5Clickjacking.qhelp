<!DOCTYPE qhelp PUBLIC
 "-//Semmle//qhelp//EN"
 "qhelp.dtd">
<qhelp>
  <overview>
    <p>
    </p>
  </overview>

  <recommendation>
    <p>
      <!-- Set frameOptions to `"deny"` either through
	   `window["sap-ui-options"]` or `data-sap-ui-frameOptions` -->
    </p>
  </recommendation>

  <example>
    <p>
      <!-- 1. Sample UI5 application that has `frameOptions` set to `"allow"` -->
    </p>
    <sample src="FrameOptionsSetToAllow.js" />
    <p>
      <!-- 2. Sample UI5 application that has no `frameOptions` set -->
    </p>
    <sample src="FrameOptionsUseDefault.js" />
    <p>
      <!-- Sample UI5 application that explicitly sets frameOptions to `"deny"` -->
    </p>
    <sample src="FrameOptionsSetToDeny.js" />
  </example>

  <references>
    <li>
      OWASP:
      <a href="https://cheatsheetseries.owasp.org/cheatsheets/Clickjacking_Defense_Cheat_Sheet.html">Clickjacking Defense Cheat Sheet</a>.
    </li>
    <li>
      Mozilla:
      <a href="https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/X-Frame-Options">X-Frame-Options</a>
    </li>
    <li>
      SAP:
      <a href="https://sapui5.hana.ondemand.com/sdk/#/topic/62d9c4d8f5ad49aa914624af9551beb7.html">Frame Options</a>
    </li>
    <li>
      SAP:
      <a href="https://sapui5.hana.ondemand.com/sdk/#/topic/d04a6d41480c4396af16b5d2b25509ec.html">Allowlist Service</a>
    </li>
  </references>
</qhelp>
