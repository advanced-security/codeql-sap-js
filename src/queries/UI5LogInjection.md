# Log Injection

A UI5 application may leak sensitive information through its logging facility if it directly or indirectly passes the data around which ends up in a logging function provided by UI5. It is also possible that a malicious actor submits a crafted user input which might lead to forging log entries.

## Recommendation

The author of the application should be aware of what sensitive information there is, prepare appropriate sanitizers such as anonymizing functions, and keep check of how and whether the logged information is being sanitized.

Also, avoid directly logging untrusted input from a remote source. Only log parts of the input, not the whole content, so that it becomes indistinguishable from system-generated log messages.

## Examples

This UI5 application directly outputs what the user submitted via the `sap.m.Input` control.

``` xml
<sap.ui.core.mvc.View controllerName="vulnerable.controller.app">
  <SearchField id="searchTodoItemsInput"
  width="20rem" search=".onSearchCompleted" /> <!-- Source -->
</sap.ui.core.mvc.View>
```

``` javascript
sap.ui.define(
  [
    "sap/ui/core/mvc/Controller",
    "sap/ui/model/json/JSONModel",
    "sap/base/Log/info",
  ],
  function (Controller, JSONModel, info) {
    return Controller.extend("vulnerable.controller.app", {
      onSearchCompleted: function () {
        var oView = this.getView();
        var oSearchField = oView.byId("searchTodoItemsInput");
        var searchValue = oSearchField.getValue();
        info(searchValue); // Sink
      },
    });
  },
);

## References

- OWASP: [Log Injection](https://www.owasp.org/index.php/Log_Injection)
- SAP: [namespace `sap/base/Log`](https://sapui5.hana.ondemand.com/sdk/#api/module:sap/base/Log)
