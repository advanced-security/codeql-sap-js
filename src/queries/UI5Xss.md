# Client-side cross-site scripting

Receiving text from the user, most notably through a control, and rendering it as HTML in another control can lead to a cross-site scripting vulnerability.

## Recommendation

### Preventing XSS Involving User Defined Control
 If the XSS attack vector includes a user-defined control, then we can mitigate the issue by improving on the implementation of the control. Specifically, we can change the method calls on instances of `sap.ui.core.RenderManager` (henceforth `RenderManager`).

- When using the now-deprecated older API with `RenderManager.write` or `RenderManager.writeAttribute`, use their respective counterparts `RenderManager.writeEscaped` and `RenderManager.writeAttributeEscaped` which sanitizes their rendered contents.
- When using the newer API  with `apiVersion: 2` (dubbed Semantic Rendering), do not use `RenderManager.unsafeHtml` unless the control property `sanitizeContent` is set to `true`.
- Regardless of the API version, it is also a good idea to use escaping functions in `sap.base.security`. Relevant sanitizers include `encodeXML` and `encodeHTML`.
- Lastly, enforce the property to something other than `string` or `any`. If a value should be used, then opt for the `enum` type which only allows a predefined set of strings.

### Preventing XSS Not Involving User Defined Control

An XSS attack vector can still exist even when no user-defined control is used. In this case, a model property or a control property act as a intermediate step when external data is passed. The view component of a UI5 application can be declared such that one control can funnel its property value to that of another.

In this case, the UI5 application should not use the property as is, but should sanitize the contents before reading it. Such sanitization can take place in the controller or in the view declaration using expression bindings.

## Example

### Custom Control with Custom Rendering Method

This custom control `vulnerable.control.xss` calls `unsafeHtml` on a given `RenderManager` instance in its static renderer function. Since its `text` property is an unrestricted string type, it can point to a string with contents that can be interpreted as HTML. If it is the case, `unsafeHtml` will render the string, running a possibly embedded JavaScript code in it.

```javascript
sap.ui.define(["sap/ui/core/Control"], function (Control) {
    return Control.extend("vulnerable.control.xss", {
        metadata: { properties: { text: { type: "string" } } },
        renderer: {
            apiVersion: 2,
            render: function (oRm, oControl) {
                oRm.openStart("div", oControl);
                oRm.unsafeHtml(oControl.getText()); // sink
                oRm.close("div");
            }
        }
    });
})
```

This is the same custom control without the possibility of XSS using several means of sanitization: The property `text` is enforced to a non-string type, hence disallows unrestricted strings. Also, the `sap.base.security.encodeXML` function is used to escape HTML control characters.

```javascript
sap.ui.define(["sap/ui/core/Control", "sap/base/security/encodeXML"], function (Control, encodeXML) {
    return Control.extend("vulnerable.control.xss", {
        metadata: { properties: { text: { type: "int" } } }, // 1. constrain the type
        renderer: {
            apiVersion: 2,
            render: function (oRm, oControl) {
                oRm.openStart("div", oControl);
                oRm.unsafeHtml(encodeXML(oControl.getText()); // 2. use security functions
            
    oRm.close("div");
            }
        }
    });
})
```

### Library Control

This example contains only library controls that are not user-defined. The untrusted user input flows from `sap.m.Input` and directly flows out via `sap.ui.core.HTML` through the model property `input` as declared in the `onInit` method of the controller.

``` xml
<sap.ui.core.mvc.View controllerName="vulnerable.controller.app">
  <sap.m.Input value="{/input}" />	 <!-- XSS Source -->
  <sap.ui.core.HTML content="{/input}"/> <!-- XSS Sink -->
</sap.ui.core.mvc.View>
```

``` javascript
sap.ui.define(["sap/ui/core/mvc/Controller", "sap/ui/model/json/JSONModel"],
  function (Controller, JSONModel) {
    return Controller.extend("vulnerable.controller.app", {
      onInit: function () {
        var oData = { input: null };
        var oModel = new JSONModel(oData);
        this.getView().setModel(oModel);
      },
    });
  },
);
```

The issue can be resolved by using an expression binding that denotes applying a sanitizer to the data binding path, or by setting the `HTML` control's `sanitizeContent` attribute to true.

``` xml
<sap.ui.core.mvc.View controllerName="vulnerable.controller.app">
  <sap.m.Input value="{/input}" />
  <sap.ui.core.HTML conten="sap.base.security.encodeXML({/input})"
  sanitizeContent="true" />
</sap.ui.core.mvc.View>
```

## References

- OWASP: [DOM Based XSS](https://owasp.org/www-community/attacks/DOM_Based_XSS).
- SAP: [Cross-site Scripting](https://sapui5.hana.ondemand.com/sdk/#/topic/91f0bd316f4d1014b6dd926db0e91070.html) in UI5.
- SAP: [Prevention of Cross-site Scripting](https://sapui5.hana.ondemand.com/sdk/#/topic/4de64e2e191f4a7297d4fd2d1e233a2d.html) in UI5.
- SAP: [API Documentation of sap.ui.core.RenderManager](https://sapui5.hana.ondemand.com/sdk/#/api/sap.ui.core.RenderManager)
- SAP: [Documentation of sap.ui.core.HTML](https://sapui5.hana.ondemand.com/sdk/#/api/sap.ui.core.HTML%23methods/setSanitizeContent)
- SAP: [Defining Control Properties](https://sapui5.hana.ondemand.com/sdk/#/topic/ac56d92162ed47ff858fdf1ce26c18c4.html).
- SAP: [Expression Binding](https://sapui5.hana.ondemand.com/sdk/#/topic/daf6852a04b44d118963968a1239d2c0).
- Common Weakness Enumeration: [CWE-79](https://cwe.mitre.org/data/definitions/79.html).
- Common Weakness Enumeration: [CWE-116](https://cwe.mitre.org/data/definitions/116.html).
