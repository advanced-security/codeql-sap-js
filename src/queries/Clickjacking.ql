/**
 * @name Clickjacking
 * @description The absence of frame options allows for clickjacking.
 * @kind problem
 * @problem.severity error
 * @security-severity 6.1
 * @precision low
 * @id js/ui5-clickjacking
 * @tags security
 *      external/cwe/cwe-451
 */

import javascript
import models.UI5HTML

from FrameOptions frameOptions
where frameOptions.allowsAllOriginEmbedding() or thereIsNoFrameOptionSet()
select frameOptions, "Clickjacking vulnerability due to $@.", frameOptions, frameOptions.toString()
