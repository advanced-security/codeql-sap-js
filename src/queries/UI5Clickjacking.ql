/**
 * @name Clickjacking
 * @description The absence of frame options allows for clickjacking.
 * @kind problem
 * @problem.severity error
 * @security-severity 6.1
 * @precision low
 * @id js/ui5-clickjacking
 * @tags security
 *      external/cwe/cwe-451
 */

import javascript
import models.UI5HTML
import semmle.javascript.RestrictedLocations
private import models.UI5

class HtmlStartTag extends HTML::DocumentElement, FirstLineOf {
  HtmlStartTag() {
    exists(UI5::Project p | this.getFile().(FirstLineOf).getFile() = p.getMainHTML())
  }
}

newtype TAlertLocation =
  TFrameOptions(FrameOptions frameOptions) or
  THtmlStartTag(HtmlStartTag htmlStartTag)

class AlertLocation extends TAlertLocation {
  FrameOptions asFrameOptions() { this = TFrameOptions(result) }

  HtmlStartTag asHtmlStartTag() { this = THtmlStartTag(result) }

  string toString() {
    result = this.asFrameOptions().toString() or
    result = this.asHtmlStartTag().toString()
  }

  predicate hasLocationInfo(string path, int sl, int sc, int el, int ec) {
    this.asFrameOptions().getLocation().hasLocationInfo(path, sl, sc, el, ec)
    or
    this.asHtmlStartTag().hasLocationInfo(path, sl, sc, el, ec)
  }
}

from AlertLocation alert, string message
where
  exists(FrameOptions frameOptions | frameOptions.allowsAllOriginEmbedding() |
    alert.asFrameOptions() = frameOptions and
    message =
      "Possible clickjacking vulnerability due to " + frameOptions.toString() +
        " being set to `allow`."
  )
  or
  exists(UI5::Project p | thereIsNoFrameOptionSet(p) |
    alert.asHtmlStartTag().getFile() = p.getMainHTML() and
    message = "Possible clickjacking vulnerability due to missing frame options."
  )
select alert, message
