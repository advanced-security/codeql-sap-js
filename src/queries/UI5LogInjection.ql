/**
 * @name Log injection
 * @description Building log entries from user-controlled sources is vulnerable to
 *              insertion of forged log entries by a malicious user.
 * @kind path-problem
 * @problem.severity error
 * @security-severity 7.8
 * @precision medium
 * @id js/ui5-log-injection
 * @tags security
 *       external/cwe/cwe-117
 */

import javascript
import models.UI5DataFlowShared
import models.UI5DataFlowShared::UI5PathGraph
import semmle.javascript.security.dataflow.LogInjectionQuery as LogInjection

class UI5LogInjectionConfiguration extends LogInjection::LogInjectionConfiguration {
  override predicate isAdditionalFlowStep(
    DataFlow::Node start, DataFlow::Node end, DataFlow::FlowLabel inLabel,
    DataFlow::FlowLabel outLabel
  ) {
    super.isAdditionalFlowStep(start, end, inLabel, outLabel)
    or
    UI5DataFlow::isAdditionalFlowStep(start, end, inLabel, outLabel)
  }
}

/**
 * An remote source associated with a `UI5BoundNode`
 */
class UI5ModelSource extends UI5DataFlow::UI5ModelSource, LogInjection::Source { }

/**
 * A log-injection sink associated with a `UI5BoundNode`
 */
class UI5ModelSink extends UI5DataFlow::UI5ModelSink, LogInjection::Sink { }

from
  UI5LogInjectionConfiguration cfg, UI5PathGraph::UI5PathNode source,
  UI5PathGraph::UI5PathNode sink, UI5PathGraph::UI5PathNode primarySource,
  UI5PathGraph::UI5PathNode primarySink
where
  cfg.hasFlowPath(source.asDataFlowPathNode(), sink.asDataFlowPathNode()) and
  primarySource = source.getAPrimarySource() and
  primarySink = sink.getAPrimarySink()
select primarySink, primarySource, primarySink, "Log entry depends on a $@.", primarySource,
  "user-provided value"
