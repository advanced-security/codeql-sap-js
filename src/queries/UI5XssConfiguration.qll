import javascript
import models.UI5::UI5
import models.UI5View
import models.UI5AMDModule
import semmle.javascript.security.dataflow.DomBasedXssQuery as DomBasedXss
import DataFlow::PathGraph

class UI5XssConfiguration extends DomBasedXss::Configuration {
  /**
   * Additional Flow Step:
   * Binding path in the model <-> control metadata
   */
  private predicate bidiModelControl(DataFlow::Node start, DataFlow::Node end) {
    exists(Project p, DataFlow::SourceNode property, Metadata metadata, UI5BoundNode node |
      // same project
      p.isInThisProject(metadata.getFile()) and
      p.isInThisProject(node.getFile()) and
      // same control
      metadata.getControl().getName() = node.getBindingPath().getControlName() and
      // restrict to interesting property types
      property.getAPropertySource("type").getStringValue() = ["string"] and
      property = metadata.getProperty(node.getBindingPath().getPropertyName()) and
      (
        start = property and end = node
        or
        start = node and end = property
      )
    )
  }

  override predicate isAdditionalFlowStep(
    DataFlow::Node start, DataFlow::Node end, DataFlow::FlowLabel inLabel,
    DataFlow::FlowLabel outLabel
  ) {
    inLabel = "taint" and
    outLabel = "taint" and
    (
      bidiModelControl(start, end)
      or
      /* 1. Control metadata property being the intermediate flow node */
      exists(string propName, Metadata metadata |
        // getAWrite -> control metadata
        start = metadata.getAWrite(propName).getArgument(1) and
        end = metadata.getProperty(propName)
        or
        // control metadata -> getTitle
        start = metadata.getProperty(propName) and
        end = metadata.getARead(propName)
      )
      or
      /* 2. Model property being the intermediate flow node */
      // JS object property (corresponding to binding path) -> getProperty('/path')
      exists(UI5BoundNode p, GetBoundValue getP |
        start = p and
        end = getP and
        p = getP.getBind()
      )
      or
      // setProperty('/path') -> JS object property (corresponding to binding path)
      exists(UI5BoundNode p, SetBoundValue setP |
        start = setP and
        end = p and
        p = setP.getBind()
      )
      // or
      /* 3. Argument to JSONModel constructor being the intermediate flow node */
      // exists(UI5 model, GetBoundValue getP |
      //   start = getP and
      //   model.getPathString() = getP.getArgument(0).asExpr().(StringLiteral).getValue() and
      //   end = model.(JsonModel).getAnArgument() and
      //   end.asExpr() instanceof StringLiteral
      // )
    )
  }
}

/**
 * Models dataflow nodes bound to a UI5 View via binding path
 */
class UI5BoundNode extends DataFlow::Node {
  UI5BindingPath bindingPath;

  UI5BindingPath getBindingPath() { result = bindingPath }

  UI5BoundNode() {
    /* The relevant portion of the content of a JSONModel */
    exists(Property p, JsonModel model, Project project |
      // The property bound to an UI5View source
      this.(DataFlow::PropRef).getPropertyNameExpr() = p.getNameExpr() and
      // The binding path refers to this model
      bindingPath.getAbsolutePath() = model.getPathString(p) and
      project.isInThisProject(this.getFile()) and
      project.isInThisProject(bindingPath.getFile())
    )
    or
    /* The URI string to the JSONModel constructor call */
    exists(JsonModel model, Project project |
      this = model.getArgument(0) and
      this.asExpr() instanceof StringLiteral and
      bindingPath.getAbsolutePath() = model.getPathString() and
      project.isInThisProject(this.getFile()) and
      project.isInThisProject(bindingPath.getFile())
    )
  }
}

/**
 * An remote source associated with a `UI5BoundNode`
 */
class UI5ModelSource extends UI5BoundNode, DomBasedXss::Source {
  UI5ModelSource() { bindingPath = any(UI5View view).getASource() }
}

/**
 * An html injection sink associated with a `UI5BoundNode`
 */
class UI5ModelSink extends UI5BoundNode, DomBasedXss::Sink {
  UI5View view;

  UI5ModelSink() {
    not view.getController().getModel().(JsonModel).isOneWayBinding() and
    bindingPath = view.getAnHtmlISink()
  }
}

/**
 * Models calls to `Model.getProperty` and `Model.getObject`
 */
class GetBoundValue extends DataFlow::CallNode {
  UI5BoundNode bind;

  GetBoundValue() {
    // direct read access to a binding path
    this.getCalleeName() = ["getProperty", "getObject"] and
    bind.getBindingPath().getAbsolutePath() = this.getArgument(0).getStringValue() and
    bind.getBindingPath().getModel() = this.getReceiver().getALocalSource()
  }

  UI5BoundNode getBind() { result = bind }
}

/**
 * Models calls to `Model.setProperty` and `Model.setObject`
 */
class SetBoundValue extends DataFlow::Node {
  UI5BoundNode bind;

  SetBoundValue() {
    exists(DataFlow::CallNode setProp |
      // direct access to a binding path
      this = setProp.getArgument(1) and
      setProp.getCalleeName() = ["setProperty", "setObject"] and
      bind.getBindingPath().getAbsolutePath() = setProp.getArgument(0).getStringValue() and
      bind.getBindingPath().getModel() = setProp.getReceiver().getALocalSource()
    )
  }

  UI5BoundNode getBind() { result = bind }
}

Locatable getUI5SourceLocation(DataFlow::Node node, string bindingPathStr) {
  result = node.(UI5ModelSource).getBindingPath() and
  result = any(UI5View view).getASource() and
  bindingPathStr = node.(UI5ModelSource).getBindingPath().getAbsolutePath()
  or
  result = node.asExpr() and
  not node.asExpr() instanceof StringLiteral and // exception on JSONModel's URI argument
  bindingPathStr = node.toString()
}

Locatable getUI5SinkLocation(DataFlow::Node node, string bindingPathStr) {
  result = node.(UI5ModelSink).getBindingPath() and
  result = any(UI5View view).getAnHtmlISink() and
  bindingPathStr = node.(UI5ModelSource).getBindingPath().getAbsolutePath()
  or
  result = node.asExpr() and
  not node.asExpr() instanceof StringLiteral and // exception on JSONModel's URI argument
  bindingPathStr = node.toString()
}
