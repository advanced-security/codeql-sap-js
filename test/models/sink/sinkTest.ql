/**
 * @id xss-sinks
 * @name XSS sinks
 * @kind problem
 */

import javascript
import queries.UI5XssConfiguration

from DataFlow::Configuration cfg, DataFlow::Node sink
where cfg.isSink(sink, _)
select getUI5SinkLocation(sink), sink.toString()
