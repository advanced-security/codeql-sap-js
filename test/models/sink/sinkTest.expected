nodes
edges
#select
| sink.js:12:31:12:34 | code | code |
| sink.js:15:52:15:55 | code | code |
| sink.js:17:24:17:27 | code | code |
| sink.js:19:25:19:28 | code | code |
| sink.js:23:22:23:25 | code | code |
| sink.js:27:18:27:21 | code | code |
| sink.js:29:27:29:30 | code | code |
| sink.js:31:34:31:37 | code | code |
| sink.js:33:21:33:24 | code | code |
| sink.js:35:21:35:24 | code | code |
| sink.js:37:23:37:26 | code | code |
